cranelift-filetests/src/function_runner.rs

```diff
@@ -75,14 +311,41 @@ impl FunctionRunner {
             );
         };
 
+        let has_check = self.options.expected.is_some() && self.options.operator.is_some();
+
         let code_page = code_page.make_exec().map_err(|e| e.to_string())?;
-        let callable_fn: fn() -> bool = unsafe { mem::transmute(code_page.as_ptr()) };
+        let ret_value_type = context.func.signature.returns.first().unwrap().value_type;
+        if !has_check && !self.options.print {
+            if !ret_value_type.is_bool() {
+                return Err(String::from("Function return type is not boolean"));
+            }
+            return match Self::invoke::<bool>(&code_page) {
+                true => Ok(()),
+                false => Err(format!("Failed: {}", context.func.name.to_string())),
+            };
+        }
 
-        // execute
-        if callable_fn() {
-            Ok(())
-        } else {
-            Err(format!("Failed: {}", context.func.name.to_string()))
+        match has_check {
+            true => Self::check(
+                code_page,
+                ret_value_type,
+                self.options.operator.unwrap(),
+                self.options.expected.unwrap(),
+                self.options.print,
+            ),
+            false => match self.options.args.len() {
+                0 => Self::check(
+                    code_page,
+                    ret_value_type,
+                    Operator::EQ,
+                    u128::from(true),
+                    self.options.print,
+                ),
+                _ => Err(format!(
+                    "Function arguments more than supported count: {}",
+                    self.options.args.len()
+                )),
+            },
```

Comment:
This was a bit hard to read but I think I get it now; it might be more clear with an enum like I mentioned above:
```
match action {
  Print => println!(...),
  Test => check(...),
  TestWithResult => check(...),
}
```

Reply:
This simplified things a lot!